### PR TITLE
Implements single Client as the default GCP SecretManager approach

### LINKED ISSUE #818: Sometimes multiple errors occur and secret files are not synchronized properly in GKE
## Version / env

- ESO `v0.4.4`
- GKE `v1.21.6-gke.1500`
- use `Workload Identity`
- use `SecretStore` (don't use ClusterSecretStore)
- use `Secret Manager`

## Problem

When I kubectl describe external-secret, I get the following error

```
Events:
Type Reason AgeFromMessage
---- ------ ---- -----------
Normal Updated56m (x415 over 7h56m)external-secretsUpdated Secret
WarningUpdateFailed 49m (x2 over 53m)external-secrets could not get secret data from provider: key "development-exsec-my-credential" from ExternalSecret "my-external-secrets-operator": unable to access Secret from SecretManager Client: rpc error: code = PermissionDenied desc = Permission 'secretmanager.versions.access' denied for resource 'projects/MY_PROJECT/secrets/development-exsec-my-credential/versions/latest' (or it may not exist).
WarningUpdateFailed 43m (x12 over 53m) external-secrets could not get secret data from provider: key "development-exsec-my-session-key" from ExternalSecret "my-external-secrets-operator": unable to access Secret from SecretManager Client: rpc error: code = Canceled desc = grpc: the client connection is closing
WarningUpdateFailed 38m external-secrets could not get secret data from provider: key "development-exsec-my-api-key" from ExternalSecret "my-external-secrets-operator": unable to access Secret from SecretManager Client: rpc error: code = Canceled desc = grpc: the client connection is closing
WarningUpdateFailed 24m (x19 over 53m) external-secrets could not get secret data from provider: key "development-exsec-my-credential" from ExternalSecret "my-external-secrets-operator": unable to access Secret from SecretManager Client: rpc error: code = Canceled desc = grpc: the client connection is closing
WarningInvalidProviderClientConfig17m (x33 over 54m) external-secrets failed to create GCP secretmanager client: unable to generate gcp access token: rpc error: code = Canceled desc = grpc: the client connection is closing
Normal Updated12m (x14 over 45m) external-secretsUpdated Secret
WarningUpdateFailed 2m42s (x10 over 34m) external-secrets could not get secret data from provider: key "development-exsec-my-bugsnag-api-key" from ExternalSecret "my-external-secrets-operator": unable to access Secret from SecretManager Client: rpc error: code = Canceled desc = grpc: the client connection is closing

```

In summary, the following types of errors occur randomly.

```
msg: could not update Secret
err: could not get secret data from provider: key \\"development-exsec-admin-bugsnag-api-key-for-api\\" from ExternalSecret \\"admin-secrets\\": unable to access Secret from SecretManager Client: rpc error: code = Canceled desc = grpc: the client connection is closing

msg: could not update Secret
err: could not get secret data from provider: key "development-exsec-my-credential" from ExternalSecret "my-external-secrets-operator": unable to access Secret from SecretManager Client: rpc error: code = PermissionDenied desc = Permission 'secretmanager.versions.access' denied for resource 'projects/MY_PROJECT/secrets/development-exsec-my-credential/versions/latest' (or it may not exist).

msg: could not get provider client
err: failed to create GCP secretmanager client: unable to generate gcp access token: rpc error: code = Canceled desc = grpc: the client connection is closing

msg: could not close provider client
err: failed to create GCP secretmanager client: unable to generate gcp access token: rpc error: code = Canceled desc = grpc: the client connection is closing

```

Lack of authorization appears to be the problem, but the following are in place.

- Workload Identity settings (NodePool/GKE) have been done.
- GSA - KSA ties are in place.
- IAM role (Workload Identity/TokenCreator) is granted.

I believe that the privileges granted are not wrong because "most of the time the Secret is created successfully, but sometimes it fails.

By the way, here is the ESO Pod log. (Ignore the fact that it does not correspond to the above error.)

<details><summary> ESO Pod log (Click me)</summary>

```json
{"level":"error","ts":1646971718.1400776,"logger":"controllers.ExternalSecret","msg":"could not close provider client","ExternalSecret":"app-dev-test1/my-secrets","SecretStore":"app-dev-test1/my-external-secrets-operator","error":"unable to close SecretManager client: rpc error: code = Canceled desc = grpc: the client connection is closing (and 4 other errors)","stacktrace":"github.com/external-secrets/external-secrets/pkg/controllers/externalsecret.(*Reconciler).Reconcile\\n\\t/home/runner/work/external-secrets/external-secrets/pkg/controllers/externalsecret/externalsecret_controller.go:246\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1646971718.6454322,"logger":"controllers.ExternalSecret","msg":"reconciled secret","ExternalSecret":"app-dev-test1/my-secrets","SecretStore":"app-dev-test1/my-external-secrets-operator"}external-secrets-c4bc855d7-9wnlb {"level":"error","ts":1646971718.7890575,"logger":"controllers.ExternalSecret","msg":"could not get provider client","ExternalSecret":"app-dev-outgame-sandbox/admin-secrets","SecretStore":"app-dev-outgame-sandbox/admin-external-secrets-operator","error":"failed to create GCP secretmanager client: unable to generate gcp access token: rpc error: code = Canceled desc = grpc: the client connection is closing","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227"}
{"level":"error","ts":1646971721.8899071,"logger":"controllers.ExternalSecret","msg":"could not update Secret","ExternalSecret":"app-dev-test2/admin-secrets","SecretStore":"app-dev-test2/admin-external-secrets-operator","error":"could not get secret data from provider: key \\"development-exsec-admin-bugsnag-api-key-for-api\\" from ExternalSecret \\"admin-secrets\\": unable to access Secret from SecretManager Client: rpc error: code = Canceled desc = grpc: the client connection is closing","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227"}
{"level":"error","ts":1646971721.890896,"logger":"controllers.ExternalSecret","msg":"could not get provider client","ExternalSecret":"prometheus/prometheus-iap-client","SecretStore":"prometheus/external-secrets-operator","error":"failed to create GCP secretmanager client: unable to generate gcp access token: rpc error: code = Canceled desc = grpc: the client connection is closing","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1646971722.2506986,"logger":"controllers.ExternalSecret","msg":"reconciled secret","ExternalSecret":"prometheus/prometheus-iap-client","SecretStore":"prometheus/external-secrets-operator"}
{"level":"error","ts":1646971781.9834816,"logger":"controllers.ExternalSecret","msg":"could not update Secret","ExternalSecret":"app-dev-test3/admin-secrets","SecretStore":"app-dev-test3/admin-external-secrets-operator","error":"could not get secret data from provider: key \\"development-exsec-admin-bugsnag-api-key-for-frontend\\" from ExternalSecret \\"admin-secrets\\": unable to access Secret from SecretManager Client: rpc error: code = PermissionDenied desc = Permission 'secretmanager.versions.access' denied for resource 'projects/MY_PROJECT/secrets/development-exsec-admin-bugsnag-api-key-for-frontend/versions/latest' (or it may not exist).","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227"}
{"level":"error","ts":1646971810.505019,"logger":"controllers.ExternalSecret","msg":"could not update Secret","ExternalSecret":"app-dev-test4/admin-secrets","SecretStore":"app-dev-test4/admin-external-secrets-operator","error":"could not get secret data from provider: key \\"development-exsec-admin-bugsnag-api-key-for-frontend\\" from ExternalSecret \\"admin-secrets\\": unable to access Secret from SecretManager Client: rpc error: code = PermissionDenied desc = Permission 'secretmanager.versions.access' denied for resource 'projects/MY_PROJECT/secrets/development-exsec-admin-bugsnag-api-key-for-frontend/versions/latest' (or it may not exist).","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227"}

```

</details>

## Concerns

We have two GKE clusters running on one GCP project, both running ESO.
Tentatively named `dev cluster` / `staging cluster`, this problem only occurs in the `dev cluster`.

The ESO startup configuration for the `dev cluster` and `staging cluster` is exactly the same.
The only difference is that the `dev cluster` has many `external secrets` and `secret stores` running because it is a development environment.

```
# dev cluster
❯ k get -A externalsecrets.external-secrets.io| wc -l
68
❯ k get -A secretstores.external-secrets.io | wc -l
47

```

```
# staging cluster
❯ k get -A externalsecrets.external-secrets.io| wc -l
 9
❯ k get -A secretstores.external-secrets.io | wc -l
 5

```

## Our manifests

<details><summary> our manifests (Click me)</summary>

```yaml
apiVersion: external-secrets.io/v1alpha1
kind: ExternalSecret
metadata:
  labels:
    app.kubernetes.io/component: admin-api
    app.kubernetes.io/name: admin-api
  name: admin-iap-client-secrets
  namespace: app-dev-test0
spec:
  data:
  - remoteRef:
      key: development-exsec-admin-iap-client
      property: client_id
      version: latest
    secretKey: client_id
  - remoteRef:
      key: development-exsec-admin-iap-client
      property: client_secret
      version: latest
    secretKey: client_secret
  refreshInterval: 1m
  secretStoreRef:
    kind: SecretStore
    name: admin-external-secrets-operator
---
apiVersion: external-secrets.io/v1alpha1
kind: ExternalSecret
metadata:
  labels:
    app.kubernetes.io/component: admin-api
    app.kubernetes.io/name: admin-api
  name: admin-secrets
  namespace: app-dev-test0
spec:
  data:
  - remoteRef:
      key: development-exsec-admin-bugsnag-api-key-for-api
      version: latest
    secretKey: bugsnag-api-key-for-api
  - remoteRef:
      key: development-exsec-admin-bugsnag-api-key-for-frontend
      version: latest
    secretKey: bugsnag-api-key-for-frontend
  refreshInterval: 1m
  secretStoreRef:
    kind: SecretStore
    name: admin-external-secrets-operator
  target:
    creationPolicy: Owner
    name: admin-secrets
---
apiVersion: external-secrets.io/v1alpha1
kind: ExternalSecret
metadata:
  labels:
    app.kubernetes.io/component: api
    app.kubernetes.io/name: my-api
  name: my-secrets
  namespace: app-dev-test0
spec:
  data:
  - remoteRef:
      key: development-exsec-my-credential
      version: latest
    secretKey: credentials.json
  - remoteRef:
      key: development-exsec-my-session-key
      version: latest
    secretKey: session-key
  - remoteRef:
      key: development-exsec-my-api-key
      version: latest
    secretKey: api-key
  - remoteRef:
      key: development-exsec-my-bugsnag-api-key
      version: latest
    secretKey: bugsnag-api-key
  refreshInterval: 1m
  secretStoreRef:
    kind: SecretStore
    name: my-external-secrets-operator
  target:
    creationPolicy: Owner
    name: my-secrets
```

```yaml
apiVersion: external-secrets.io/v1alpha1
kind: SecretStore
metadata:
  labels:
    app.kubernetes.io/component: admin-api
    app.kubernetes.io/name: admin-api
  name: admin-external-secrets-operator
  namespace: app-dev-test0
spec:
  provider:
    gcpsm:
      auth:
        workloadIdentity:
          clusterLocation: asia-northeast1
          clusterName: development
          serviceAccountRef:
            name: admin-external-secrets-operator
      projectID: MY_PROJECT
---
apiVersion: external-secrets.io/v1alpha1
kind: SecretStore
metadata:
  labels:
    app.kubernetes.io/component: api
    app.kubernetes.io/name: my-api
  name: my-external-secrets-operator
  namespace: app-dev-test0
spec:
  provider:
    gcpsm:
      auth:
        workloadIdentity:
          clusterLocation: asia-northeast1
          clusterName: development
          serviceAccountRef:
            name: my-external-secrets-operator
      projectID: MY_PROJECT
```

</details>

## What we tried

I tried and checked the following, but no improvement.

- Set `concurrent=10` to ESO in `dev-cluster`.
- Set `.spec.refreshInterval` to `5m` in `externalsecrets` used in `dev-cluster` (originally `1m`)
- Check GCP Quota for Secret Manager API / IAM API (`staging-cluster` is running in the same PJ so this should not be a problem)

I also took the debug log but it doesn't seem to have any useful information.

<details><summary> debug log(Click me)</summary>

```
2022-03-11T06:41:03.937506827Z {"level":"debug","ts":1646980863.9366648,"logger":"controllers.ExternalSecret","msg":"reconciled secret","ExternalSecret":"app-dev-test0/admin-secrets","SecretStore":"app-dev-test0/admin-external-secrets-operator"}
2022-03-11T06:41:03.937595199Z {"level":"debug","ts":1646980863.9367306,"logger":"events","msg":"Normal","object":{"kind":"ExternalSecret","namespace":"app-dev-test0","name":"admin-secrets","uid":"5bdb1a3b-226b-41d3-b116-eaf5e55cc0c7","apiVersion":"external-secrets.io/v1alpha1","resourceVersion":"370114762"},"reason":"Updated","message":"Updated Secret"}
2022-03-11T06:41:04.059129393Z {"level":"debug","ts":1646980864.0589173,"logger":"controllers.ExternalSecret","msg":"skipping refresh","ExternalSecret":"app-dev-test0/admin-secrets","SecretStore":"app-dev-test0/admin-external-secrets-operator","rv":"1-472f358594e87ba4678b952d5f0d02af"}
2022-03-11T06:41:30.059628782Z {"level":"error","ts":1646980890.059401,"logger":"controllers.ExternalSecret","msg":"could not get provider client","ExternalSecret":"app-dev-test0/admin-iap-client-secrets","SecretStore":"app-dev-test0/admin-external-secrets-operator","error":"failed to create GCP secretmanager client: unable to generate gcp access token: rpc error: code = Canceled desc = grpc: the client connection is closing","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227"}
2022-03-11T06:41:30.060736443Z {"level":"debug","ts":1646980890.0605001,"logger":"events","msg":"Warning","object":{"kind":"ExternalSecret","namespace":"app-dev-test0","name":"admin-iap-client-secrets","uid":"500f9d9f-8d68-4e18-804b-9df866414a54","apiVersion":"external-secrets.io/v1alpha1","resourceVersion":"370112794"},"reason":"InvalidProviderClientConfig","message":"failed to create GCP secretmanager client: unable to generate gcp access token: rpc error: code = Canceled desc = grpc: the client connection is closing"}
2022-03-11T06:41:30.385601122Z {"level":"info","ts":1646980890.3854601,"logger":"controllers.ExternalSecret","msg":"reconciled secret","ExternalSecret":"app-dev-test0/admin-iap-client-secrets","SecretStore":"app-dev-test0/admin-external-secrets-operator"}
2022-03-11T06:41:30.386757849Z {"level":"debug","ts":1646980890.3862054,"logger":"events","msg":"Normal","object":{"kind":"ExternalSecret","namespace":"app-dev-test0","name":"admin-iap-client-secrets","uid":"500f9d9f-8d68-4e18-804b-9df866414a54","apiVersion":"external-secrets.io/v1alpha1","resourceVersion":"370115999"},"reason":"Updated","message":"Updated Secret"}
2022-03-11T06:41:30.494851945Z {"level":"debug","ts":1646980890.4946878,"logger":"controllers.ExternalSecret","msg":"skipping refresh","ExternalSecret":"app-dev-test0/admin-iap-client-secrets","SecretStore":"app-dev-test0/admin-external-secrets-operator","rv":"2-aa72504d829be8d0dc98bffd91af7175"}
2022-03-11T06:41:30.496088871Z {"level":"error","ts":1646980890.495925,"logger":"controllers.ExternalSecret","msg":"could not close provider client","ExternalSecret":"app-dev-test0/admin-iap-client-secrets","SecretStore":"app-dev-test0/admin-external-secrets-operator","error":"unable to close SecretManager client: rpc error: code = Canceled desc = grpc: the client connection is closing (and 4 other errors)","stacktrace":"github.com/external-secrets/external-secrets/pkg/controllers/externalsecret.(*Reconciler).Reconcile\\n\\t/home/runner/work/external-secrets/external-secrets/pkg/controllers/externalsecret/externalsecret_controller.go:189\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266\\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\\n\\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227"}
2022-03-11T06:42:00.188171213Z {"level":"debug","ts":1646980920.1880045,"logger":"controllers.ExternalSecret","msg":"skipping refresh","ExternalSecret":"app-dev-test0/admin-iap-client-secrets","SecretStore":"app-dev-test0/admin-external-secrets-operator","rv":"2-aa72504d829be8d0dc98bffd91af7175"}
2022-03-11T06:42:03.827023557Z {"level":"debug","ts":1646980923.826863,"logger":"controllers.ExternalSecret","msg":"reconciled secret","ExternalSecret":"app-dev-test0/my-secrets","SecretStore":"app-dev-test0/my-external-secrets-operator"}
2022-03-11T06:42:03.829087766Z {"level":"debug","ts":1646980923.8288572,"logger":"events","msg":"Normal","object":{"kind":"ExternalSecret","namespace":"app-dev-test0","name":"my-secrets","uid":"78d5e757-064c-4214-981b-98889d1af799","apiVersion":"external-secrets.io/v1alpha1","resourceVersion":"370113132"},"reason":"Updated","message":"Updated Secret"}
2022-03-11T06:42:03.946847552Z {"level":"debug","ts":1646980923.9466808,"logger":"controllers.ExternalSecret","msg":"skipping refresh","ExternalSecret":"app-dev-test0/my-secrets","SecretStore":"app-dev-test0/my-external-secrets-operator","rv":"2-b8c4fa48ab589f778dd09a8f2b0ea8e7"}
2022-03-11T06:42:04.268476009Z {"level":"debug","ts":1646980924.2669353,"logger":"controllers.ExternalSecret","msg":"reconciled secret","ExternalSecret":"app-dev-test0/admin-secrets","SecretStore":"app-dev-test0/admin-external-secrets-operator"}
2022-03-11T06:42:04.268531595Z {"level":"debug","ts":1646980924.2683086,"logger":"events","msg":"Normal","object":{"kind":"ExternalSecret","namespace":"app-dev-test0","name":"admin-secrets","uid":"5bdb1a3b-226b-41d3-b116-eaf5e55cc0c7","apiVersion":"external-secrets.io/v1alpha1","resourceVersion":"370115711"},"reason":"Updated","message":"Updated Secret"}
2022-03-11T06:42:04.380377285Z {"level":"debug","ts":1646980924.3801599,"logger":"controllers.ExternalSecret","msg":"skipping refresh","ExternalSecret":"app-dev-test0/admin-secrets","SecretStore":"app-dev-test0/admin-external-secrets-operator","rv":"1-472f358594e87ba4678b952d5f0d02af"}
2022-03-11T06:43:03.803177998Z {"level":"debug","ts":1646980983.8029563,"logger":"controllers.SecretStore","msg":"validating","secretstore":"app-dev-test0/admin-external-secrets-operator"}
2022-03-11T06:43:03.901338841Z {"level":"debug","ts":1646980983.9011183,"logger":"events","msg":"Normal","object":{"kind":"SecretStore","namespace":"app-dev-test0","name":"admin-external-secrets-operator","uid":"c1e70348-046b-441d-9a25-1e49d9e113fb","apiVersion":"external-secrets.io/v1alpha1","resourceVersion":"370107419"},"reason":"Valid","message":"store validated"}
2022-03-11T06:43:04.602448342Z {"level":"debug","ts":1646980984.6022263,"logger":"controllers.ExternalSecret","msg":"reconciled secret","ExternalSecret":"app-dev-test0/admin-secrets","SecretStore":"app-dev-test0/admin-external-secrets-operator"}
2022-03-11T06:43:04.604262051Z {"level":"debug","ts":1646980984.6022446,"logger":"events","msg":"Normal","object":{"kind":"ExternalSecret","namespace":"app-dev-test0","name":"admin-secrets","uid":"5bdb1a3b-226b-41d3-b116-eaf5e55cc0c7","apiVersion":"external-secrets.io/v1alpha1","resourceVersion":"370116354"},"reason":"Updated","message":"Updated Secret"}
2022-03-11T06:43:04.716556753Z {"level":"debug","ts":1646980984.7163868,"logger":"controllers.ExternalSecret","msg":"skipping refresh","ExternalSecret":"app-dev-test0/admin-secrets","SecretStore":"app-dev-test0/admin-external-secrets-operator","rv":"1-472f358594e87ba4678b952d5f0d02af"}
2022-03-11T06:43:05.331476139Z {"level":"debug","ts":1646980985.3313165,"logger":"controllers.SecretStore","msg":"validating","secretstore":"app-dev-test0/my-external-secrets-operator"}
2022-03-11T06:43:05.439749825Z {"level":"debug","ts":1646980985.4394543,"logger":"events","msg":"Normal","object":{"kind":"SecretStore","namespace":"app-dev-test0","name":"my-external-secrets-operator","uid":"cff1fcff-67f6-4e43-b629-e24c86ae2c9b","apiVersion":"external-secrets.io/v1alpha1","resourceVersion":"370086544"},"reason":"Valid","message":"store validated"}

```

</details>

### LINKED ISSUE #834: High pod memory consumption
# Problem 

We recently started using ESO with GKE in my organization. I’ve been monitoring the memory consumption that the pods have and, I’ve noticed that it continuously increases its consumption without a reason.  Due to this I’ve started to limit how much memory it can consume but this causes the pod to restart when it reaches the memory limit and restarts the whole process again.
image

I wanted to see if someone could help me understand this behavior since at least from a logs perspective nothing seems to be happening.

```logs
{"level":"info","ts":1647279723.910821,"logger":"controller.secretstore","msg":"Starting EventSource","reconciler group":"external-secrets.io","reconciler kind":"SecretStore","source":"kind source: *v1alpha1.SecretStore"}
{"level":"info","ts":1647279723.910864,"logger":"controller.secretstore","msg":"Starting Controller","reconciler group":"external-secrets.io","reconciler kind":"SecretStore"}
{"level":"info","ts":1647279723.9109325,"logger":"controller.externalsecret","msg":"Starting EventSource","reconciler group":"external-secrets.io","reconciler kind":"ExternalSecret","source":"kind source: *v1alpha1.ExternalSecret"}
{"level":"info","ts":1647279723.9109786,"logger":"controller.externalsecret","msg":"Starting EventSource","reconciler group":"external-secrets.io","reconciler kind":"ExternalSecret","source":"kind source: *v1.Secret"}
{"level":"info","ts":1647279723.9109404,"logger":"controller.clustersecretstore","msg":"Starting EventSource","reconciler group":"external-secrets.io","reconciler kind":"ClusterSecretStore","source":"kind source: *v1alpha1.ClusterSecretStore"}
{"level":"info","ts":1647279723.9110146,"logger":"controller.clustersecretstore","msg":"Starting Controller","reconciler group":"external-secrets.io","reconciler kind":"ClusterSecretStore"}
{"level":"info","ts":1647279723.910985,"logger":"controller.externalsecret","msg":"Starting Controller","reconciler group":"external-secrets.io","reconciler kind":"ExternalSecret"}
{"level":"info","ts":1647279724.0114942,"logger":"controller.secretstore","msg":"Starting workers","reconciler group":"external-secrets.io","reconciler kind":"SecretStore","worker count":1}
{"level":"info","ts":1647279724.0116224,"logger":"controller.clustersecretstore","msg":"Starting workers","reconciler group":"external-secrets.io","reconciler kind":"ClusterSecretStore","worker count":1}
{"level":"info","ts":1647279724.0117655,"logger":"controller.externalsecret","msg":"Starting workers","reconciler group":"external-secrets.io","reconciler kind":"ExternalSecret","worker count":1}
```

Initially, we thought that the `refreshInterval` could be a cause for all of those but we have changed our settings to use a specific version and with a `refreshInterval: "0"` to avoid refresh but no luck.

## Environment
- GKE: `1.21.9-gke.1001`
- ESO:  `0.3.11`, `0.4.2`, `0.4.4`  <- We have used these versions to see if it is fixed but the behavior is present on all.
- Workload Identity
- GCP Secret Manager 

## Deployment

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: external-secrets
spec:
  replicas: 1
  selector:
    matchLabels:
      app: externalsecrets
  template:
    spec:
    metadata:
      labels:
        app: externalsecrets
      serviceAccountName: external-secrets
      containers:
        - name: external-secrets
          image: ghcr.io/external-secrets/external-secrets:v0.4.2
          imagePullPolicy: IfNotPresent
          args:
          - --concurrent=1
          ports:
            - containerPort: 8080
              protocol: TCP
          resources:
            limits:
              cpu: 100m
              memory: 128Mi
            requests:
              cpu: 100m
              memory: 128Mi
```

### LINKED ISSUE #837: Make sure Validate method is implemented throughout providers
Providers missing Validate Methods:
- [x] #782
- [x] #886
- [x] #887
- [x] #888
- [x] #889
- [x] #890
- [x] #891
- [x] #892
- [x] #893

### PR DESCRIPTION
**Pull request by gusfcarvalho**
Partially covers #818
 Fixes #834 

Some comments around this fix:

The issues are around the fact that apparently GCP SM does not expects multiple clients running at the same time. 

Also, the main memory leak comes from the fact that SecretStore reconciles were not calling the Close method.

The main idea of the fix is to make sure we only have one instance of GCPSM and GCPWI clients, and never close them. Closing doesn't have to do with memory per se, but with some random error messages that appear if a given ES or SS gets reconciled in parallel.

Signed-off-by: Gustavo Carvalho <gustavo.carvalho@container-solutions.com>

### CODE REVIEW

**Comment by gusfcarvalho** [comment]
@moolen Alternatively to this fix (which is provider - related), what I already tested and confirmed to work is implement a mutex approach. This would obviously impact the concurrency option, but it seems for some providers this can actually make sense. So the other way around this is to add a new `SupportsConcurrency` method for the Providers and a check on the reconciler loop before Locking a given Mutex.

I followed with this one as the fix as this way concurrency is still allowed.


PS - I believe I still can do some cleanup to methods that are not used anymore, but I want to see what e2e tests will say about this.

**Comment by gusfcarvalho** [comment]
Also, a picture about the improvement:

fix running with concurrency=10, store-requeue-interval="1s"

main running with concurrency=1, store-requeue-interval="10s"

(hided graphs are for other combinations of these two parameters)

image

**Comment by knelasevero** [comment]
Awesome! Also curious to see the mutex implementation, if it is already there. We can decide which direction to go looking at them side by side

**Comment by gusfcarvalho** [comment]
I did fix-by-mutex on top of 0.4.4 only, but it can be checked in here: https://github.com/external-secrets/external-secrets/tree/fix-by-mutex

**Comment by gusfcarvalho** [comment]
Update: I just figured out that even if we do this approach, GCP provider itself currently wouldn't support multiple concurrent calls. This is the issue of #818 actually. Basically what's happening there is that one thread is configuring to use one credentials, while the other is trying to fetch secrets. At this time, the actual call the client does is using the newly configured credentials by thread 1, and not the expected in thread2. This, when granular permissions are used, turns out to be a Permission Denied error.

TL;DR - It's safer even from a security point of view to implement Mutexes, and just not allow GCP Provider to be concurrent. I'll close this PR and add a new one with this config.

The new PR will also be from a fork, as requested by @knelasevero =)
